src/libcore/str/pattern.rs

```diff
@@ -630,3 +509,420 @@ impl<'a, F> Pattern<'a> for F where F: FnMut(char) -> bool {
+                self.position += needle.len();
+                if !long_period {
+                    self.memory = 0;
+                }
+                continue 'search;
+            }
+
+            // See if the right part of the needle matches
+            let start = if long_period { self.crit_pos }
+                        else { cmp::max(self.crit_pos, self.memory) };
+            for i in start..needle.len() {
+                if needle[i] != haystack[self.position + i] {
+                    self.position += i - self.crit_pos + 1;
+                    if !long_period {
+                        self.memory = 0;
+                    }
+                    continue 'search;
+                }
+            }
+
+            // See if the left part of the needle matches
+            let start = if long_period { 0 } else { self.memory };
+            for i in (start..self.crit_pos).rev() {
+                if needle[i] != haystack[self.position + i] {
+                    self.position += self.period;
+                    if !long_period {
+                        self.memory = needle.len() - self.period;
+                    }
+                    continue 'search;
+                }
+            }
+
+            // We have found a match!
+            let match_pos = self.position;
+            self.position += needle.len(); // add self.period for all matches
+            if !long_period {
+                self.memory = 0; // set to needle.len() - self.period for all matches
+            }
+
+            return SearchStep::Match(match_pos, match_pos + needle.len());
+        }
+    }
+
+    // Follows the ideas in `next()`.
+    //
+    // Note: No documentation has been used for the algorithm in `next_back()`,
+    // simply trying to adapt forward search to backward search with extensive
+    // testing.
+    #[inline]
+    fn next_back(&mut self, haystack: &[u8], needle: &[u8]) -> SearchStep {
+        // `next_back()` uses `self.end` as its cursor -- so that `next()` and `next_back()`
+        // are independent.
+        let start = self.end;
+        'search: loop {
+            // Check that we have room to search in
+            if needle.len() > self.end {
+                if start == 0 {
+                    return SearchStep::Done
+                } else {
+                    self.end = 0;
+                }
+            }
+            if start != self.end {
+                return SearchStep::Reject(self.end, start);
+            }
+
+            // Quickly skip by large portions unrelated to our substring
+            if (self.byteset >>
+                    ((haystack[self.end - needle.len()] & 0x3f)
+                     as usize)) & 1 == 0 {
+                self.end -= needle.len();
+                continue 'search;
+            }
+
+            // See if the left part of the needle matches
+            let start = self.crit_pos;
+            for (i, &b) in needle[..start].iter().enumerate().rev() {
```

Comment:
Could either this be changed to an iteration over a range or the loop in the forward search be changed to using `.iter().enumerate()`? It would be nice to keep the two sides consistent.